DetDescr/src/DetDescr/EcalGeometry.cxx

```diff
@@ -629,4 +629,54 @@ bool EcalGeometry::isInside(double normX, double normY) const {
   return (dotProd > 0.);
 }
 
+bool EcalGeometry::isFiducialInModule(double x, double y, int layer_id) const {
+  // now assume we know the layer
+  // shift to center of layer
+  // and convert to flower coordinates
+  double p{x - std::get<0>(layer_pos_xy_.at(layer_id))},
+      q{y - std::get<1>(layer_pos_xy_.at(layer_id))};
+
+  // deduce module ID
+  //    there are only 7 modules so we just loop through them
+  //    all and pick out the module ID that we are inside of
+
+  int module_id{-1};
+  for (auto const& [mid, module_xy] : module_pos_xy_) {
+    double probe_x{p - module_xy.first}, probe_y{q - module_xy.second};
+    if (cornersSideUp_) rotate(probe_x, probe_y);
+    if (isInside(probe_x / moduleR_, probe_y / moduleR_)) {
+      module_id = mid;
+      break;
+    }
+  }
+
+  if (module_id < 0) {
+    return false;
+  } else {
+    return true;
+  }
+}
+
+bool EcalGeometry::isFiducialInCell(double x, double y, int layer_id,
+                                    int module_id) const {
+  // now assume we know the layer and module
+  // shift to center of layer and then center of module
+  double p{x - std::get<0>(layer_pos_xy_.at(layer_id)) -
+           module_pos_xy_.at(module_id).first},
+      q{y - std::get<1>(layer_pos_xy_.at(layer_id)) -
+        module_pos_xy_.at(module_id).second};
+
+  // need to rotate
+  if (cornersSideUp_) rotate(p, q);
+
+  // deduce cell ID
+  int cell_id = cell_id_in_module_.FindBin(p, q) - 1;
+
+  if (cell_id < 0) {
+    return false;
+  } else {
+    return true;
+  }
+}
+
```

Comment:
I am interested in reducing duplication so that this rather complicated logic is eaiser to maintain. My proposal is to add a `fallible` flag to the `getID` cascade of functions. If `fallible` is `false`, then the current behavior of throwing exceptions is done. This is helpful for SimCore where we want any `getID` call not being found to be treated as a run-ending bug in the geometry. If `fallible` is `true`, then an ID is returned that needs to be checked for validity. We could use the null ID `ldmx::EcalID(0)` (checked with `id.null()`) as the "invalid" ID or [std::optional](https://en.cppreference.com/w/cpp/utility/optional).